app/controllers/carts_controller.rb

```diff
@@ -1,8 +1,24 @@
 # frozen_string_literal: true
 
 class CartsController < ApplicationController
-  before_action :authenticate_user!
+  before_action :authenticate_user!, :set_cart
   def show
-    render :show, locals: { items: current_user.cart.cart_items }
+    render :show, locals: { items: @cart.cart_items }
+  end
+
+  def add_product
+    product = Product.find(params[:id])
+    add_product = CartServices::AddProduct.new.call @cart, product
+    if add_product.success?
+      redirect_to root_path, notice: 'Product added successfully'
+    else
+      redirect_back fallback_location: root_path, alert: add_product.payload[:error]
+    end
+  end
+
+  private
+
+  def set_cart
+    @cart = current_user.cart
   end
```

Comment:
Kłuje w oczy ta małpa, kiedy wszędzie mam zmienne lokalne. Może coś takiego?
```suggestion
  def cart
    current_user.cart
  end
```
i wywalić to z before_action ?

Reply:
tak wywal - callbacki to zlo. Ja lubie zrobic wlasnie cos jak proponujesz:
```
  def cart
    @_cart ||= current_user.cart
  end
```